DatePicker: add attribute defaultTime

nighca opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide ([中文](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.zh-CN.md) | [English](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.en-US.md) | [Español](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.es.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer relative issues for you PR.

Details: #7646

nighca commented:
@wacky6 @Leopoldthecoder Plz review. And, should I update document files in this PR?

wacky6 commented:
CI failed for some reasons. Can you check it?

wacky6 commented:
Yes, you can update the document. Just write English or Chinese (or both), the Spanish one will be translated afterwards.

I prefer a separate section explaining the two new behaviors (do not reset & reset to defaultTime). Each should have a demo.

nighca commented:
Ok, I will check it and add docs when free.

nighca commented:
@wacky6 I fixed the CI. Also I added some docs and an example for the attribute `defaultTime`. Plz review.

I've run into a strange problem that the test case fails in Travis if I use `DELAY` [here](https://github.com/ElemeFE/element/pull/9094/files#diff-a9901ef39f8a888687ce5f8e4b8fd47bR1289), while it passes on my own computer. I have to use a longer delay time(`DELAY * 2`) to make it work in Travis environment.

Can you plz check if it is fine to do so, or if there is any better solution.

wacky6 commented:
The cause of `DELAY * 2` is strange. `createVue` / `createTest` should not require additional time to complete. In fact all other tests work well without `*2`.

Also, there are improvements I would like to address / discuss. See PR reviews.

nighca commented:
@wacky6 I can not find reviews or comments. Have you submitted them?

wacky6 commented:
View on github. You should see comments in changes.

nighca commented:
@wacky6 No comments in the changed tab, either. If there are reviews, they will appear among our conversations.

Leopoldthecoder commented:
@wacky6 I couldn't find comments either. Could you have a check?

wacky6 commented:
Sorry. Seems I forgot to submit it.

nighca commented on examples/docs/en-US/date-picker.md:
```diff
@@ -394,6 +444,7 @@ This feature is at alpha stage. Feedback welcome.
 | picker-options | additional options, check the table below | object | — | {} |
 | range-separator | range separator | string | — | '-' |
 | default-value | optional, default date of the calendar | Date | anything accepted by `new Date()` | — |
+| default-time | optional, the time value to use when select datetime range in date table (type `datetimerange`) | Date[] | Array with length 1 or 2, each item accepted by `new Date()`. The first item for the start datetime and then second item for the end datetime | — |
```
Agree.

nighca commented on test/unit/specs/date-picker.spec.js:
```diff
@@ -1246,6 +1246,93 @@ describe('DatePicker', () => {
+
+      const vmWithDefaultTime = createTest(DatePicker, {
+        type: 'datetimerange',
+        value: [new Date(2000, 10, 10, 10, 10), new Date(2000, 10, 11, 10, 10)],
```
OK.

nighca commented on examples/docs/en-US/date-picker.md:
```diff
@@ -64,7 +64,11 @@
+        value11: '',
+        value12: '',
+        value13: '',
+        defaultTimeWithMin: [new Date(2018, 1, 1, 12)],
```
It will be easier for users to provide time string instead of `Date` parsable object. But it will introduce a new format of time data and parse behavior, which means extra API & realization complexity. I'm not sure if it deserves that.

I found another problem of this PR: it is impossible to configure millisecond value of `defaultTime` now (because I'm using `modifyTime`). I will fix that.

wacky6 replied:
I think fecha's parse method can handle time alone. (dateUtil in 'date-picker/util/index.js')
You can also refer to time-select panel's parseTime.

wacky6 replied:
I don't think millisecond value is an issue since it is not respected in any picker panels (They may reset millisecond to 0).

Also I don't think there are compelling use cases for it. For example, using half-open intervals link `[00:00:00.000, +1 days 00:00:00.000)` instead of closed interval `[00:00:00.000, 23:59:59.999]` eliminates the need of millisecond values.

nighca replied:
Thanks, I'll check it.

nighca commented:
@wacky6 Fixed.

wacky6 commented:
LGTM, DEMO: https://jsfiddle.net/kxmnmLej/

TODOs:
- [ ] How about "do not reset time if null/undefined" as mentioned in https://github.com/ElemeFE/element/issues/7646#issuecomment-353934059
- [ ] (after merge) es doc

Leopoldthecoder commented:
@nighca Thanks for contributing.

Actually `default-time` also affects pickers with `type="daterange"` (not only "on the date panel with type `datetimerange`"), which is a good thing, so after merging I'll update the docs to reflect this.

Are we good to merge this PR now? @wacky6

wacky6 commented:
Yes, can merge.
Can finish Todos afterwards

Leopoldthecoder commented:
@nighca This PR breaks the hovering background color when picking a range:

> ![untitled](https://user-images.githubusercontent.com/10095631/34989367-3b9baf42-fafd-11e7-80e1-2bb71d8e259c.gif)


The correct behavior is:
> ![untitled1](https://user-images.githubusercontent.com/10095631/34989384-48375a44-fafd-11e7-8b80-367d52546e6f.gif)

Could you take another look?

nighca commented:
@Leopoldthecoder Sure.

nighca commented:
@Leopoldthecoder @wacky6 I created another PR to fix this, in which I also added some assertions to test this behavior. Plz review.
